File: resolution.py
from itertools import combinations

def negate(literal):
    return literal[1:] if literal.startswith("¬") else "¬" + literal

def resolvent(c1, c2, lit):
    nlit = negate(lit)
    if lit in c1 and nlit in c2:
        return (c1 - {lit} | c2 - {nlit})
    elif nlit in c1 and lit in c2:
        return (c1 - {nlit} | c2 - {lit})
    raise ValueError(f"Literal '{lit}' kommt nicht komplementär in {c1} und {c2} vor")

def all_resolvents(c1, c2, known_clauses):
    steps = []
    new_resolvents = set()
    for lit in c1:
        nlit = negate(lit)
        if nlit in c2:
            res = frozenset(resolvent(c1, c2, lit))
            if res in known_clauses:
                steps.append({
                    "type": "cut_known",
                    "c1": list(c1), "c2": list(c2),
                    "literal": lit,
                    "resolvent": list(res)
                })
            else:
                steps.append({
                    "type": "cut_new",
                    "c1": list(c1), "c2": list(c2),
                    "literal": lit,
                    "resolvent": list(res)
                })
                new_resolvents.add(res)
        else:
            steps.append({
                "type": "no_cut",
                "c1": list(c1), "c2": list(c2),
                "literal": lit
            })
    return steps, new_resolvents

def resolution_level(clauses, known_clauses):
    level_steps = []
    new_resolvents = set()
    for c1, c2 in combinations(clauses, 2):
        steps, new_res = all_resolvents(c1, c2, known_clauses)
        level_steps.extend(steps)
        new_resolvents |= new_res
    return level_steps, new_resolvents

def full_resolution(initial_clauses):
    all_clauses = set(initial_clauses)
    levels = []
    level_num = 1
    
    while True:
        level_steps, new_resolvents = resolution_level(all_clauses, all_clauses)
        levels.append({
            "level": level_num,
            "clauses": [list(c) for c in all_clauses],
            "steps": level_steps,
            "new_resolvents": [list(r) for r in new_resolvents]
        })
        
        if not new_resolvents:
            break
        
        all_clauses |= new_resolvents
        level_num += 1
    
    return levels

File: frontend/resolution.js
function negate(literal) {
    return literal.startsWith('!') ? literal.slice(1) : '!' + literal;
}

function resolvent(c1, c2, lit) {
    const nlit = negate(lit);
    if (c1.includes(lit) && c2.includes(nlit)) {
        const res = [...c1.filter(x => x !== lit), ...c2.filter(x => x !== nlit)];
        return Array.from(new Set(res)).sort();
    } else if (c1.includes(nlit) && c2.includes(lit)) {
        const res = [...c1.filter(x => x !== nlit), ...c2.filter(x => x !== lit)];
        return Array.from(new Set(res)).sort();
    }
    throw new Error(`Literal '${lit}' kommt nicht komplementär in ${c1} und ${c2} vor`);
}

function allResolvents(c1, c2, knownClauses, foundInThisLevel) {
    const steps = [];
    const newResolvents = [];
    for (const lit of c1) {
        const nlit = negate(lit);
        if (c2.includes(nlit)) {
            const res = resolvent(c1, c2, lit);
            if (knownClauses.some(cl => arraysEqual(cl, res)) || foundInThisLevel.some(cl => arraysEqual(cl, res))) {
                steps.push({ type: 'cut_known', c1, c2, literal: lit, resolvent: res });
            } else {
                steps.push({ type: 'cut_new', c1, c2, literal: lit, resolvent: res });
                newResolvents.push(res);
            }
        } else {
            steps.push({ type: 'no_cut', c1, c2, literal: lit });
        }
    }
    return [steps, newResolvents];
}

function resolutionLevel(clauses, knownClauses) {
    const levelSteps = [];
    const newResolvents = [];
    const foundInThisLevel = [];

    for (let i = 0; i < clauses.length; i++) {
        for (let j = i + 1; j < clauses.length; j++) {
            const [steps, newRes] = allResolvents(clauses[i], clauses[j], knownClauses, foundInThisLevel);
            levelSteps.push(...steps);
            for (const r of newRes) {
                if (!newResolvents.some(cl => arraysEqual(cl, r))) {
                    newResolvents.push(r);
                    foundInThisLevel.push(r);
                }
            }
        }
    }
    return [levelSteps, newResolvents];
}

function fullResolution(initialClauses) {
    const allClauses = initialClauses.map(cl => [...cl]);
    const levels = [];
    let levelNum = 1;

    while (true) {
        const [levelSteps, newResolvents] = resolutionLevel(allClauses, allClauses);
        levels.push({ level: levelNum, clauses: allClauses.map(cl => [...cl]), steps: levelSteps, new_resolvents: newResolvents.map(cl => [...cl]) });

        if (newResolvents.length === 0) break;

        for (const r of newResolvents) {
            if (!allClauses.some(cl => arraysEqual(cl, r))) allClauses.push(r);
        }
        levelNum++;
    }
    return levels;
}

// Hilfsfunktion, um Array-Gleichheit zu prüfen
function arraysEqual(a, b) {
    if (a.length !== b.length) return false;
    for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) return false;
    return true;
}

// resolution.js
export { fullResolution, negate, resolvent, allResolvents };
